File: scripts/parse_stat_blocks.py
"""
Parse D&D Beyond HTML stat blocks (from JSON array of HTML strings) into
stat-block JSON objects matching the SRD schema.
"""

import json
import re
import sys
from pathlib import Path
from typing import Optional


def unesc(s: str) -> str:
    s = s.strip()
    if s.startswith('"'):
        s = s[1:]
    if s.endswith('"'):
        s = s[:-1]
    return s.replace("\\n", "\n").replace('\\"', '"')


def strip_html(html: str) -> str:
    return re.sub(r"<[^>]+>", " ", html).strip()


def parse_size_type_align(text: str) -> tuple[Optional[str], Optional[str], Optional[str], Optional[str]]:
    """Return (size, creatureType, descriptiveTags, alignment)."""
    if not text or not text.strip():
        return (None, None, None, None)
    # e.g. "Large Aberration, Lawful Evil" or "Medium or Small Humanoid, Neutral Evil"
    # or "Medium Fiend (Yugoloth), Neutral Evil"
    size_part = r"((?:Tiny|Small|Medium|Large|Huge|Gargantuan)(?:\s+or\s+(?:Tiny|Small|Medium|Large|Huge|Gargantuan))?)"
    match = re.match(
        rf"^{size_part}\s+(.+?),\s*(.+)$",
        text.strip(),
    )
    if not match:
        return (None, None, None, None)
    size = match.group(1).strip()
    type_and_tags = match.group(2).strip()
    alignment = match.group(3).strip()
    # type_and_tags e.g. "Aberration" or "Fiend (Yugoloth)"
    paren = type_and_tags.find("(")
    if paren >= 0:
        creature_type = type_and_tags[:paren].strip()
        descriptive_tags = type_and_tags[paren:].strip()
    else:
        creature_type = type_and_tags
        descriptive_tags = None
    return (size, creature_type, descriptive_tags, alignment)


def parse_skills(raw: Optional[str]) -> Optional[list[str]]:
    """Split 'Perception +4 , Persuasion +4 , Stealth +7' into ['Perception +4', 'Persuasion +4', 'Stealth +7']."""
    if not raw or not raw.strip():
        return None
    return [part.strip() for part in raw.split(",") if part.strip()]


def parse_senses(raw: Optional[str]) -> Optional[list[str]]:
    """Split 'Darkvision 60 ft.; Passive Perception 14' into ['Darkvision 60 ft.', 'Passive Perception 14']."""
    if not raw or not raw.strip():
        return None
    return [part.strip() for part in raw.split(";") if part.strip()]


def parse_comma_separated_list(raw: Optional[str]) -> list[str]:
    """Split 'Cold, Fire, Lightning' into ['Cold', 'Fire', 'Lightning']. Returns [] when raw is None or empty."""
    if not raw or not raw.strip():
        return []
    return [part.strip() for part in raw.split(",") if part.strip()]


def parse_speed(raw: Optional[str]) -> tuple[Optional[str], Optional[str], Optional[str], Optional[str], Optional[str]]:
    """Return (speed, burrowSpeed, climbSpeed, flySpeed, swimSpeed)."""
    if not raw or not raw.strip():
        return (None, None, None, None, None)
    text = raw.strip()
    burrow = None
    climb = None
    fly = None
    swim = None
    # Match "Burrow 20 ft.", "Climb 30 ft.", "Fly 30 ft. (hover)", "Swim 40 ft."
    # Allow period in value so we get "30 ft." not "30 ft"
    burrow_m = re.search(r"Burrow\s+([^,;]+?)(?=\s*[,;]|$)", text, re.I)
    if burrow_m:
        burrow = burrow_m.group(1).strip()
    climb_m = re.search(r"Climb\s+([^,;]+?)(?=\s*[,;]|$)", text, re.I)
    if climb_m:
        climb = climb_m.group(1).strip()
    fly_m = re.search(r"Fly\s+([^,;]+?)(?=\s*[,;]|$)", text, re.I)
    if fly_m:
        fly = fly_m.group(1).strip()
    swim_m = re.search(r"Swim\s+([^,;]+?)(?=\s*[,;]|$)", text, re.I)
    if swim_m:
        swim = swim_m.group(1).strip()
    # Walk speed is the first segment (before first comma or the whole thing)
    first = text.split(",")[0].strip()
    # If first segment is a plain speed (e.g. "30 ft.") and not a keyword speed, it's walk
    if not re.match(r"^(Burrow|Climb|Fly|Swim)\s", first, re.I):
        speed = first
    else:
        speed = None
    return (speed, burrow, climb, fly, swim)


def parse_cr_line(text: str) -> tuple[Optional[str], Optional[int], Optional[str], Optional[int]]:
    """Return (challengeRating, experiencePoints, experiencePointsAlt, proficiencyBonus)."""
    if not text or not text.strip():
        return (None, None, None, None)
    text = text.strip()
    # e.g. "10 (XP 5,900, or 7,200 in lair; PB +4 )"
    cr_match = re.match(r"^(\d+(?:/\d+)?)\s*\(?", text)
    challenge_rating = cr_match.group(1) if cr_match else None
    xp_match = re.search(r"XP\s*([\d,]+)", text)
    experience_points = int(xp_match.group(1).replace(",", "")) if xp_match else None
    alt_match = re.search(r"or\s+([\d,]+)\s+in\s+lair", text, re.I)
    experience_points_alt = alt_match.group(1).strip() if alt_match else None
    pb_match = re.search(r"PB\s*([+-]?\d+)", text)
    proficiency_bonus = int(pb_match.group(1)) if pb_match else None
    return (challenge_rating, experience_points, experience_points_alt, proficiency_bonus)


def parse_abilities(html: str) -> dict[str, dict[str, int]]:
    out: dict[str, dict[str, int]] = {}
    for ab in ["Str", "Dex", "Con", "Int", "Wis", "Cha"]:
        pat = (
            r"<th>"
            + re.escape(ab)
            + r"</th>\s*<td>(\d+)</td>(?:\s*<[^>]+>)*\s*<strong>([+-]?\d+)</strong>(?:\s*<[^>]+>)*\s*<strong>([+-]?\d+)</strong>"
        )
        m = re.search(pat, html, re.DOTALL)
        if m:
            out[ab] = {
                "score": int(m.group(1)),
                "mod": int(m.group(2)),
                "save": int(m.group(3)),
            }
    return out


def parse_one(html_raw: str) -> dict:
    s = unesc(html_raw)
    texts: list[str] = []
    for p in re.findall(r"<p[^>]*>(.*?)</p>", s, re.DOTALL):
        texts.append(re.sub(r"\s+", " ", strip_html(p)).strip())

    name = None
    m = re.search(
        r'<a[^>]*class="[^"]*monster-tooltip[^"]*"[^>]*>([^<]+)</a>', s
    )
    if m:
        name = m.group(1).strip()

    size_type_align = texts[0] if texts else None
    size, creature_type, descriptive_tags, alignment = parse_size_type_align(
        size_type_align or ""
    )

    ac = None
    init_mod = None
    init_score = None
    hp = None
    hp_dice = None
    speed_raw = None
    skills_raw = None
    senses_raw = None
    languages = None
    cr_line = None
    resistances_raw = None
    immunities_raw = None
    vulnerabilities = None
    gear = None

    for t in texts:
        if "AC" in t and "Initiative" in t:
            ma = re.search(r"AC\s*(\d+)", t)
            if ma:
                ac = int(ma.group(1))
            ma = re.search(r"Initiative\s*([+-]?\d+)\s*\((\d+)\)", t)
            if ma:
                init_mod = ma.group(1)
                init_score = int(ma.group(2))
        elif "HP" in t and "(" in t:
            ma = re.search(r"(\d+)\s*\(([^)]+)\)", t)
            if ma:
                hp = int(ma.group(1))
                hp_dice = ma.group(2).strip()
        elif t.startswith("Speed "):
            speed_raw = t.replace("Speed ", "", 1).strip()
        elif t.startswith("Skills "):
            skills_raw = t.replace("Skills ", "", 1).strip()
        elif t.startswith("Senses "):
            senses_raw = t.replace("Senses ", "", 1).strip()
        elif t.startswith("Languages "):
            languages = t.replace("Languages ", "", 1).strip()
        elif t.startswith("CR "):
            cr_line = t.replace("CR ", "", 1).strip()
        elif t.startswith("Resistances "):
            resistances_raw = t.replace("Resistances ", "", 1).strip()
        elif t.startswith("Immunities "):
            immunities_raw = t.replace("Immunities ", "", 1).strip()
        elif t.startswith("Vulnerabilities "):
            vulnerabilities = t.replace("Vulnerabilities ", "", 1).strip()
        elif t.startswith("Gear "):
            gear = t.replace("Gear ", "", 1).strip()

    challenge_rating, experience_points, experience_points_alt, proficiency_bonus = (
        parse_cr_line(cr_line or "")
    )

    speed, burrow_speed, climb_speed, fly_speed, swim_speed = parse_speed(speed_raw)
    skills = parse_skills(skills_raw)
    senses = parse_senses(senses_raw)
    resistances = parse_comma_separated_list(resistances_raw)
    immunities = parse_comma_separated_list(immunities_raw)

    abilities = parse_abilities(s)

    headers = list(re.finditer(r'<p class="monster-header"[^>]*>([^<]+)</p>', s))
    legendary_intro = None
    legendary_match = re.search(
        r'<p class="legendary-actions"[^>]*>(.*?)</p>', s, re.DOTALL
    )
    if legendary_match:
        legendary_intro = re.sub(r"\s+", " ", strip_html(legendary_match.group(1))).strip()

    traits: list[str] = []
    actions: list[str] = []
    bonus_actions: list[str] = []
    reactions: list[str] = []
    legendary_actions: list[str] = []

    for i, H in enumerate(headers):
        sec = H.group(1).strip()
        start = H.end()
        end = headers[i + 1].start() if i + 1 < len(headers) else len(s)
        block = s[start:end]
        paras = re.findall(r'<p data-content-chunk-id="[^"]*">(.*?)</p>', block, re.DOTALL)
        if sec == "Legendary Actions" and legendary_intro:
            legendary_actions.append(legendary_intro)
            legendary_intro = None
        for p in paras:
            t = re.sub(r"\s+", " ", strip_html(p)).strip()
            if not t:
                continue
            if sec == "Traits":
                traits.append(t)
            elif sec == "Actions":
                actions.append(t)
            elif sec == "Bonus Actions":
                bonus_actions.append(t)
            elif sec == "Reactions":
                reactions.append(t)
            elif sec == "Legendary Actions":
                legendary_actions.append(t)
    if legendary_intro and not legendary_actions:
        legendary_actions.append(legendary_intro)

    flat_abilities: dict[str, int] = {}
    ab_map = {
        "Str": "strength",
        "Dex": "dexterity",
        "Con": "constitution",
        "Int": "intelligence",
        "Wis": "wisdom",
        "Cha": "charisma",
    }
    for ab, key in ab_map.items():
        if ab in abilities:
            flat_abilities[key] = abilities[ab]["score"]
            flat_abilities[key + "Save"] = abilities[ab]["save"]

    return {
        "name": name,
        "size": size,
        "creatureType": creature_type,
        "descriptiveTags": descriptive_tags,
        "alignment": alignment,
        "armorClass": ac,
        "initiativeModifier": init_mod,
        "initiativeScore": init_score,
        "hitPoints": hp,
        "hitPointDice": hp_dice,
        "speed": speed,
        "speedBurrow": burrow_speed,
        "speedClimb": climb_speed,
        "speedFly": fly_speed,
        "speedSwim": swim_speed,
        **flat_abilities,
        "skills": skills,
        "resistances": resistances,
        "immunities": immunities,
        "vulnerabilities": vulnerabilities,
        "gear": gear,
        "senses": senses,
        "languages": languages,
        "challengeRating": challenge_rating,
        "experiencePoints": experience_points,
        "experiencePointsAlt": experience_points_alt,
        "proficiencyBonus": proficiency_bonus,
        "traits": traits,
        "actions": actions,
        "bonusActions": bonus_actions,
        "reactions": reactions,
        "legendaryActions": legendary_actions,
    }


def main() -> None:
    if len(sys.argv) != 3:
        print("Usage: parse_stat_blocks.py <input.json> <output.json>", file=sys.stderr)
        sys.exit(1)
    input_path = Path(sys.argv[1])
    output_path = Path(sys.argv[2])
    if not input_path.exists():
        print(f"Input file not found: {input_path}", file=sys.stderr)
        sys.exit(1)
    with open(input_path, encoding="utf-8") as f:
        data = json.load(f)
    if not isinstance(data, list):
        print("Input JSON must be an array of HTML strings", file=sys.stderr)
        sys.exit(1)
    parsed = [parse_one(entry) for entry in data]
    output_path.parent.mkdir(parents=True, exist_ok=True)
    with open(output_path, "w", encoding="utf-8") as f:
        json.dump(parsed, f, indent=2, ensure_ascii=False)
    print(f"Parsed {len(parsed)} stat blocks to {output_path}")


if __name__ == "__main__":
    main()

File: scripts/parseStatBlocks.ts
/* eslint-disable @typescript-eslint/no-loop-func */
/**
 * Parse D&D Beyond HTML stat blocks (from JSON array of HTML strings) into
 * stat-block JSON objects matching the SRD schema.
 *
 * Note: This was written and maintained by AI agents
 */

import { mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import { dirname } from 'node:path';
import process from 'node:process';

const SIZE_PART = /(?:Tiny|Small|Medium|Large|Huge|Gargantuan)(?:\s+or\s+(?:Tiny|Small|Medium|Large|Huge|Gargantuan))?/;
const SIZE_TYPE_ALIGN_RE = new RegExp(`^(${SIZE_PART.source})\\s+(.+?),\\s*(.+)$`);

interface AbilityEntry {
  mod: number;
  save: number;
  score: number;
}

interface ParsedStatBlock {
  actions: string[];
  alignment: null | string;
  armorClass: null | number;
  bonusActions: string[];
  challengeRating: null | string;
  charisma: number;
  charismaSave: number;
  constitution: number;
  constitutionSave: number;
  creatureType: null | string;
  descriptiveTags: null | string;
  dexterity: number;
  dexteritySave: number;
  experiencePoints: null | number;
  experiencePointsAlt: null | string;
  gear: null | string;
  hitPointDice: null | string;
  hitPoints: null | number;
  immunities: string[];
  initiativeModifier: null | string;
  initiativeScore: null | number;
  intelligence: number;
  intelligenceSave: number;
  languages: null | string;
  legendaryActions: string[];
  name: null | string;
  proficiencyBonus: null | number;
  reactions: string[];
  resistances: string[];
  senses: null | string[];
  size: null | string;
  skills: null | string[];
  speed: null | string;
  speedBurrow: null | string;
  speedClimb: null | string;
  speedFly: null | string;
  speedSwim: null | string;
  strength: number;
  strengthSave: number;
  traits: string[];
  vulnerabilities: null | string;
  wisdom: number;
  wisdomSave: number;
}

const unesc = (s: string): string => {
  let out = s.trim();
  if (out.startsWith('"')) out = out.slice(1);
  if (out.endsWith('"')) out = out.slice(0, -1);
  return out.replace(/\\n/g, '\n').replace(/\\"/g, '"');
};

const stripHtml = (html: string): string => html.replace(/<[^>]+>/g, ' ').trim();

const parseSizeTypeAlign = (text: string): [null | string, null | string, null | string, null | string] => {
  const trimmed = text.trim();

  if (trimmed === '') return [null, null, null, null];

  const match = trimmed.match(SIZE_TYPE_ALIGN_RE);

  if (!match) return [null, null, null, null];

  const [, size, typeAndTags, alignment] = match.map((m: string | undefined) => m?.trim());

  if (size == null || typeAndTags == null || alignment == null) return [null, null, null, null];

  const paren = typeAndTags.indexOf('(');
  const creatureType = paren >= 0 ? typeAndTags.slice(0, paren).trim() : typeAndTags;
  const descriptiveTags = paren >= 0 ? typeAndTags.slice(paren).trim() : null;
  return [size, creatureType, descriptiveTags, alignment];
};

const parseSkills = (raw: null | string | undefined): null | string[] => {
  if (raw === null || raw === undefined || raw.trim() === '') return null;
  return raw
    .split(',')
    .map(p => p.trim())
    .filter(p => p !== '');
};

const parseSenses = (raw: null | string | undefined): null | string[] => {
  if (raw === null || raw === undefined || raw.trim() === '') return null;
  return raw
    .split(';')
    .map(p => p.trim())
    .filter(p => p !== '');
};

const parseCommaSeparatedList = (raw: null | string | undefined): string[] => {
  if (raw === null || raw === undefined || raw.trim() === '') return [];
  return raw
    .split(',')
    .map(p => p.trim())
    .filter(p => p !== '');
};

const parseSpeed = (
  raw: null | string | undefined,
): [null | string, null | string, null | string, null | string, null | string] => {
  const text = raw?.trim() ?? '';
  if (text === '') return [null, null, null, null, null];

  const burrowM = text.match(/Burrow\s+([^,;]+?)(?=\s*[,;]|$)/i);
  const climbM = text.match(/Climb\s+([^,;]+?)(?=\s*[,;]|$)/i);
  const flyM = text.match(/Fly\s+([^,;]+?)(?=\s*[,;]|$)/i);
  const swimM = text.match(/Swim\s+([^,;]+?)(?=\s*[,;]|$)/i);
  const burrow = burrowM?.[1]?.trim() ?? null;
  const climb = climbM?.[1]?.trim() ?? null;
  const fly = flyM?.[1]?.trim() ?? null;
  const swim = swimM?.[1]?.trim() ?? null;
  const [firstSeg] = text.split(',');
  const first = firstSeg?.trim() ?? '';
  const isKeywordFirst = /^(Burrow|Climb|Fly|Swim)\s/i.test(first);
  const speed = isKeywordFirst ? null : first;
  return [speed, burrow, climb, fly, swim];
};

const parseCrLine = (text: string): [null | string, null | number, null | string, null | number] => {
  const trimmed = text.trim();

  if (trimmed === '') return [null, null, null, null];

  const crMatch = trimmed.match(/^(\d+(?:\/\d+)?)\s*\(?/);
  const challengeRating = crMatch?.[1] ?? null;
  const xpMatch = trimmed.match(/XP\s*([\d,]+)/);
  const experiencePoints = xpMatch?.[1] != null ? Number.parseInt(xpMatch[1].replace(/,/g, ''), 10) : null;
  const altMatch = trimmed.match(/or\s+([\d,]+)\s+in\s+lair/i);
  const experiencePointsAlt = altMatch?.[1]?.trim() ?? null;
  const pbMatch = trimmed.match(/PB\s*([+-]?\d+)/);
  const proficiencyBonus = pbMatch?.[1] != null ? Number.parseInt(pbMatch[1], 10) : null;

  return [challengeRating, experiencePoints, experiencePointsAlt, proficiencyBonus];
};

const ABILITY_NAMES = ['Str', 'Dex', 'Con', 'Int', 'Wis', 'Cha'] as const;

const parseAbilities = (html: string): Record<string, AbilityEntry> => {
  const out: Record<string, AbilityEntry> = {};

  for (const ab of ABILITY_NAMES) {
    const pat = new RegExp(
      `<th>${ab}</th>\\s*<td>(\\d+)</td>(?:\\s*<[^>]+>)*\\s*<strong>([+-]?\\d+)</strong>(?:\\s*<[^>]+>)*\\s*<strong>([+-]?\\d+)</strong>`,
      's',
    );
    const m = html.match(pat);
    if (m?.[1] !== undefined && m[2] !== undefined && m[3] !== undefined) {
      out[ab] = {
        mod: Number.parseInt(m[2], 10),
        save: Number.parseInt(m[3], 10),
        score: Number.parseInt(m[1], 10),
      };
    }
  }

  return out;
};

const AB_MAP: Record<string, string> = {
  Cha: 'charisma',
  Con: 'constitution',
  Dex: 'dexterity',
  Int: 'intelligence',
  Str: 'strength',
  Wis: 'wisdom',
};

const buildFlatAbilities = (abilities: Record<string, AbilityEntry>): Record<string, number> => {
  return Object.entries(AB_MAP)
    .map(([ab, key]) => [abilities[ab], key] as const)
    .filter((tuple): tuple is [AbilityEntry, string] => tuple[0] != null)
    .reduce<Record<string, number>>((acc, [entry, key]) => {
      acc[key] = entry.score;
      acc[`${key}Save`] = entry.save;
      return acc;
    }, {});
};

interface HeaderMatch {
  fullLength: number;
  index: number;
  section: string;
}

const collectHeaders = (s: string): HeaderMatch[] => {
  const headerRegex = /<p class="monster-header"[^>]*>([^<]+)<\/p>/g;
  const headers: HeaderMatch[] = [];
  let m: null | RegExpExecArray;
  while ((m = headerRegex.exec(s)) !== null) {
    const [, section] = m;
    headers.push({
      fullLength: m[0].length,
      index: m.index,
      section: section !== undefined ? section.trim() : '',
    });
  }
  return headers;
};

const extractTextsFromHtml = (s: string): string[] => {
  const pBlocks = [...s.matchAll(/<p[^>]*>([\s\S]*?)<\/p>/g)];
  return pBlocks.map(m => {
    const [, content] = m;
    if (content == null) return '';
    return stripHtml(content).replace(/\s+/g, ' ').trim();
  });
};

const extractFieldValues = (
  texts: string[],
): {
  ac: null | number;
  crLine: null | string;
  gear: null | string;
  hp: null | number;
  hpDice: null | string;
  immunitiesRaw: null | string;
  initMod: null | string;
  initScore: null | number;
  languages: null | string;
  resistancesRaw: null | string;
  sensesRaw: null | string;
  skillsRaw: null | string;
  speedRaw: null | string;
  vulnerabilities: null | string;
} => {
  let ac: null | number = null;
  let initMod: null | string = null;
  let initScore: null | number = null;
  let hp: null | number = null;
  let hpDice: null | string = null;
  let speedRaw: null | string = null;
  let skillsRaw: null | string = null;
  let sensesRaw: null | string = null;
  let languages: null | string = null;
  let crLine: null | string = null;
  let resistancesRaw: null | string = null;
  let immunitiesRaw: null | string = null;
  let vulnerabilities: null | string = null;
  let gear: null | string = null;

  for (const t of texts) {
    if (t.includes('AC') && t.includes('Initiative')) {
      const ma = t.match(/AC\s*(\d+)/);
      if (ma?.[1] !== undefined) ac = Number.parseInt(ma[1], 10);
      const initMa = t.match(/Initiative\s*([+-]?\d+)\s*\((\d+)\)/);
      if (initMa?.[1] !== undefined && initMa[2] !== undefined) {
        [, initMod] = initMa;
        initScore = Number.parseInt(initMa[2], 10);
      }
    } else if (t.includes('HP') && t.includes('(')) {
      const ma = t.match(/(\d+)\s*\(([^)]+)\)/);
      if (ma?.[1] !== undefined && ma[2] !== undefined) {
        hp = Number.parseInt(ma[1], 10);
        hpDice = ma[2].trim();
      }
    } else if (t.startsWith('Speed ')) {
      speedRaw = t.replace('Speed ', '').trim();
    } else if (t.startsWith('Skills ')) {
      skillsRaw = t.replace('Skills ', '').trim();
    } else if (t.startsWith('Senses ')) {
      sensesRaw = t.replace('Senses ', '').trim();
    } else if (t.startsWith('Languages ')) {
      languages = t.replace('Languages ', '').trim();
    } else if (t.startsWith('CR ')) {
      crLine = t.replace('CR ', '').trim();
    } else if (t.startsWith('Resistances ')) {
      resistancesRaw = t.replace('Resistances ', '').trim();
    } else if (t.startsWith('Immunities ')) {
      immunitiesRaw = t.replace('Immunities ', '').trim();
    } else if (t.startsWith('Vulnerabilities ')) {
      vulnerabilities = t.replace('Vulnerabilities ', '').trim();
    } else if (t.startsWith('Gear ')) {
      gear = t.replace('Gear ', '').trim();
    }
  }

  return {
    ac,
    crLine,
    gear,
    hp,
    hpDice,
    immunitiesRaw,
    initMod,
    initScore,
    languages,
    resistancesRaw,
    sensesRaw,
    skillsRaw,
    speedRaw,
    vulnerabilities,
  };
};

const parseSectionBlocks = (
  s: string,
  headers: HeaderMatch[],
  legendaryIntro: null | string,
): {
  actions: string[];
  bonusActions: string[];
  legendaryActions: string[];
  reactions: string[];
  traits: string[];
} => {
  const traits: string[] = [];
  const actions: string[] = [];
  const bonusActions: string[] = [];
  const reactions: string[] = [];
  let legendaryActions: string[] = [];
  let remainingLegendaryIntro = legendaryIntro;

  for (let i = 0; i < headers.length; i += 1) {
    const header = headers[i]!;
    const nextHeader = headers[i + 1];
    const sec = header.section;
    const start = header.index + header.fullLength;
    const end = nextHeader !== undefined ? nextHeader.index : s.length;
    const block = s.slice(start, end < 0 ? s.length : end);
    const paras = [...block.matchAll(/<p data-content-chunk-id="[^"]*">([\s\S]*?)<\/p>/g)];

    if (sec === 'Legendary Actions' && remainingLegendaryIntro !== null) {
      legendaryActions = [remainingLegendaryIntro];
      remainingLegendaryIntro = null;
    }

    paras
      .map(p => p[1])
      .filter(content => content != null)
      .forEach(content => {
        const t = stripHtml(content).replace(/\s+/g, ' ').trim();
        if (t !== '') {
          if (sec === 'Traits') traits.push(t);
          else if (sec === 'Actions') actions.push(t);
          else if (sec === 'Bonus Actions') bonusActions.push(t);
          else if (sec === 'Reactions') reactions.push(t);
          else if (sec === 'Legendary Actions') legendaryActions.push(t);
        }
      });
  }

  if (remainingLegendaryIntro !== null && legendaryActions.length === 0) {
    legendaryActions = [remainingLegendaryIntro];
  }

  return {
    actions,
    bonusActions,
    legendaryActions,
    reactions,
    traits,
  };
};

const defaultScores = {
  charisma: 0,
  charismaSave: 0,
  constitution: 0,
  constitutionSave: 0,
  dexterity: 0,
  dexteritySave: 0,
  intelligence: 0,
  intelligenceSave: 0,
  strength: 0,
  strengthSave: 0,
  wisdom: 0,
  wisdomSave: 0,
};

const parseOne = (htmlRaw: string): ParsedStatBlock => {
  const s = unesc(htmlRaw);
  const texts = extractTextsFromHtml(s);

  let name: null | string = null;
  const nameMatch = s.match(/<a[^>]*class="[^"]*monster-tooltip[^"]*"[^>]*>([^<]+)<\/a>/);
  if (nameMatch) {
    const [, n] = nameMatch;
    if (n !== undefined) name = n.trim();
  }

  const sizeTypeAlign = texts[0] ?? null;
  const [size, creatureType, descriptiveTags, alignment] = parseSizeTypeAlign(sizeTypeAlign ?? '');

  const fields = extractFieldValues(texts);
  const [challengeRating, experiencePoints, experiencePointsAlt, proficiencyBonus] = parseCrLine(fields.crLine ?? '');
  const [speed, burrowSpeed, climbSpeed, flySpeed, swimSpeed] = parseSpeed(fields.speedRaw);
  const skills = parseSkills(fields.skillsRaw);
  const senses = parseSenses(fields.sensesRaw);
  const resistances = parseCommaSeparatedList(fields.resistancesRaw);
  const immunities = parseCommaSeparatedList(fields.immunitiesRaw);

  const abilities = parseAbilities(s);
  const headers = collectHeaders(s);

  let legendaryIntro: null | string = null;
  const legendaryMatch = s.match(/<p class="legendary-actions"[^>]*>([\s\S]*?)<\/p>/);
  if (legendaryMatch) {
    const [_, introContent] = legendaryMatch;
    legendaryIntro = introContent !== undefined ? stripHtml(introContent).replace(/\s+/g, ' ').trim() : null;
  }

  const { actions, bonusActions, legendaryActions, reactions, traits } = parseSectionBlocks(s, headers, legendaryIntro);

  const flatAbilities = buildFlatAbilities(abilities);

  return {
    actions,
    alignment,
    armorClass: fields.ac,
    bonusActions,
    challengeRating,
    charisma: defaultScores.charisma,
    charismaSave: defaultScores.charismaSave,
    constitution: defaultScores.constitution,
    constitutionSave: defaultScores.constitutionSave,
    creatureType,
    descriptiveTags,
    dexterity: defaultScores.dexterity,
    dexteritySave: defaultScores.dexteritySave,
    experiencePoints,
    experiencePointsAlt,
    gear: fields.gear,
    hitPointDice: fields.hpDice,
    hitPoints: fields.hp,
    immunities,
    initiativeModifier: fields.initMod,
    initiativeScore: fields.initScore,
    intelligence: defaultScores.intelligence,
    intelligenceSave: defaultScores.intelligenceSave,
    languages: fields.languages,
    legendaryActions,
    name,
    proficiencyBonus,
    reactions,
    resistances,
    senses,
    size,
    skills,
    speed,
    speedBurrow: burrowSpeed,
    speedClimb: climbSpeed,
    speedFly: flySpeed,
    speedSwim: swimSpeed,
    strength: defaultScores.strength,
    strengthSave: defaultScores.strengthSave,
    traits,
    vulnerabilities: fields.vulnerabilities,
    wisdom: defaultScores.wisdom,
    wisdomSave: defaultScores.wisdomSave,
    ...flatAbilities,
  };
};

const main = (): void => {
  if (process.argv.length !== 4) {
    process.stderr.write('Usage: parseStatBlocks.ts <input.json> <output.json>\n');
    process.exit(1);
  }
  const [, , inputPath, outputPath] = process.argv;
  if (inputPath === undefined || outputPath === undefined) {
    process.stderr.write('Missing input or output path\n');
    process.exit(1);
  }

  let data: unknown;
  try {
    const raw = readFileSync(inputPath, 'utf-8');
    data = JSON.parse(raw) as unknown;
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    process.stderr.write(`Failed to read input: ${msg}\n`);
    process.exit(1);
  }

  if (!Array.isArray(data)) {
    process.stderr.write('Input JSON must be an array of HTML strings\n');
    process.exit(1);
  }

  const parsed: ParsedStatBlock[] = (data as string[]).map(parseOne);
  mkdirSync(dirname(outputPath), { recursive: true });
  writeFileSync(outputPath, JSON.stringify(parsed, null, 2), 'utf-8');
  process.stderr.write(`Parsed ${parsed.length} stat blocks to ${outputPath}\n`);
};

main();
